[EDITS] Copy file to edits folder. (#1338)
mulesoft/docs-mule-runtime-3: modules/ROOT/edits/execution-engine.adoc

> = Execution Engine

Reviewer: Is this topic really about the execution engine, or about Task-Oriented Execution?
Suggest a container title that implies or comprises the subtopics: processing types, threading, configuration, and back pressure.

ifndef::env-site,env-github[]
include::_attributes.adoc[]
endif::[]
:keywords: mule, threading, scheduler, non-blocking

Mule runtime engine implements a reactive execution engine, tuned for non-blocking and asynchronous execution. +
To see more details about reactive programming, visit https://en.wikipedia.org/wiki/Reactive_programming.

> The task-oriented execution model enables you to take advantage of non-blocking IO at high concurrency levels transparently, meaning you don’t need to account for threading or asynchronicity in your flows. Each operation inside a Mule flow is a task that provides metadata about its execution, and Mule makes tuning decisions based on that metadata.

Reviewer: The model, or this model (nonblocking and asynchronous)?

Reviewer: GLOBAL: nonblocking
GLOBAL: I/O

> Mule Event processors indicate to Mule whether they are CPU intensive, CPU light, or IO intensive operations. These workload types help Mule tune for different workloads, so you don’t need to manage thread pools manually to achieve optimum performance. Instead, Mule introspects the available resources (such as memory and CPU cores) in the system to tune thread pools automatically.

Reviewer: I looked up the presentation of "mule event processor" online and found that most references (from MuleSoft) use "Mule event processor."  Let me know if you disagree, so that I can update our Term Presentation List.

Reviewer: CPU-intensive, CPU-light, or I/O-intensive operations...

[[processing_types]]

> == Processing Types

Reviewer: ```suggestion
== Types of Mule Event Processors
```

> Mule Event processors indicate to Mule what kind of work they do, which can be one of:

Reviewer: ```suggestion
Mule event processors indicate to Mule what kind of work they do, which can be one of:
```

> * CPU Light +

Reviewer: ```suggestion
* CPU-light processor+
```

> For quick operations (around 10ms), or Non-Blocking IO, for example, a Logger (`logger`) or HTTP Request operation (`http:request`). These tasks should not perform any blocking IO. The applicable strings in the logs are `CPU_LIGHT` and `CPU_LIGHT_ASYNC`.

Reviewer: ```suggestion
For quick operations (around 10 ms), or nonblocking I/O, for example, a Logger (`logger`) or HTTP Request operation (`http:request`). These processors should not perform any blocking I/O. The applicable strings in the logs are `CPU_LIGHT` and `CPU_LIGHT_ASYNC`.
```
"These processors should not perform..." or "These operations should not include..."?

I was initially confused because I thought that this section was about types of operations (processes) rather than types of processors. It looked to me as if we were interchanging as synonyms "operation," "work," "task," and "workload type" as "processes".

> * Blocking IO +

Reviewer: ```suggestion
* Blocking I/O processors+
```

> For IO that blocks the calling thread, for example, a Database Select operation (`db:select`) or a SFTP read (`sftp:read`) . Applicable strings in the logs are `BLOCKING` and `IO`.

Reviewer: ```suggestion
For I/O that blocks the calling thread, for example, a Database Select operation (`db:select`) or an SFTP read (`sftp:read`). Applicable strings in the logs are `BLOCKING` and `IO`.
```

> * CPU Intensive +

Reviewer: ```suggestion
* CPU-intensive processors +
```

> For CPU-bound computations, usually taking more than 10ms to execute. These tasks should not perform any IO activities. Examples include the Transform Message component (`ee:transform`). The applicable string in the logs is `CPU_INTENSIVE`.

Reviewer: ```suggestion
For CPU-bound computations, usually taking more than 10 ms to execute. These processors should not perform any I/O activities. One example is the Transform Message component (`ee:transform`). The applicable string in the logs is `CPU_INTENSIVE`.
```
"These processors should not perform..." or "These operations should not include..."? Again, I'm a bit confused between processors, processes, operations, and tasks.

> See specific component or module documentation to learn the processing type it supports. If none is specified, the default is CPU Light.

Reviewer: ...the default type is CPU light.

For connectors created with the Mule SDK, the SDK determines the most
appropriate processing type based on how the connector is implemented. For
details on that mechanism, refer to the xref:1.1@mule-sdk::index.adoc[Mule SDK documentation].

[[threading]]
== Threading

Based on the processing type of a component, Mule executes that component on a thread pool that is specifically tuned for that kind of work. These thread pools are managed by Mule and shared across all apps in the same Mule instance.

> When started, Mule introspects the available resources (such as memory and CPU cores) in the system to tune thread pools automatically for the environment where it is running. The default configuration was established through performance testing, which found optimal values for most scenarios.

Reviewer: "Based on the processing type" or "the processor type"? 
(I'm still not sure whether we mean the operation or the operator: process or processor. I'll try to stop asking about this for the rest of the topic and assume that you will clarify, if necessary.)

Does "When started" refer to "Mule" (or a processor?)? Just double-checking, because that is what the syntax implies.

Reviewer: Also, does the "default configuration" refer to the ops environment; are we saying that Mule can tune thread pools based on an established environment config? 

Or is the "default configuration" the type of processor?

See if we can make the relationship between the last sentence and the one before it more clear.

Here is a minor suggestion to clarify the first sentence: 

```suggestion
When started, Mule introspects the available resources in the system (such as memory and CPU cores) so that it can automatically tune thread pools for use in that Mule instance. The default configuration was established through performance testing, which found optimal values for most scenarios.
```
Or?

```suggestion
When started, Mule introspects the available resources (such as memory and CPU cores) in the system to tune thread pools automatically for the CPU-light or CPU-intensive, or blocking I/O, environment in which Mule is running.
```

Reviewer: Should we add something here about custom thread pools?

> The different thread pools allow Mule to manage resources more efficiently, requiring significantly fewer threads (and their inherent memory footprint) to handle a given workload.

Reviewer: Which "different" thread pools; different from what? Does this work instead:
```suggestion
Tuning thread pools enables Mule to manage resources more efficiently by using only the number of threads, and their inherent memory footprint, necessary to process a given workload.
```

Reviewer: Ah! so the "different thread pools" actually refers to the different processing types.

That was not clear to me at the outset; will it be clear to your audience?

Do you suppose that we should include the description of each associated thread pool type in the initial description of the processor type, so that we don't have repeat section titles referring to CPU light (processor, thread pool), CPU intensive (processor, thread pool), and blocking I/O (processor, thread pool)?

Otherwise, consider the following slight rewordings:

> Following are described the key aspects of each thread pool.

Reviewer: ```suggestion
Thread pool types correspond to processor types: CPU light, CPU intensive, and blocking I/O.
```

> === CPU Light

Reviewer: ```suggestion
=== CPU-Light Thread Pools
```

> CPU Light is for a relatively small thread pool (2 threads per available core by default).

Reviewer: ```suggestion
A CPU-light thread pool is relatively small (two threads per available core by default).
```

> Apart from executing the CPU Light processors, this pool performs the handoff

Reviewer: ```suggestion
Apart from executing the CPU-light processors, this pool performs the handoff
```

of the event between processors in the flow (including the routers) and the

> response handling for non-blocking IO.

Reviewer: ```suggestion
response handling for nonblocking I/O.
```

Reviewer: Going to stop editing CPU-light for CPU Light; please be sure to let me know if this is really a proper noun and I will add it to the Term Presentation List. Thanks!

Sometimes, throughput might drop or become unresponsive in an app due to some code

> misusing the CPU Light thread pool. This issue can be quickly checked by taking a

Reviewer: ```suggestion
misusing the CPU Light thread pool. You can quickly resolve this issue by creating a
```

thread dump of the runtime and looking for `WAITING` or `BLOCKED` or for

> long-running processes in the CPU Light threads.

Reviewer: ...and <doing what>

> === CPU Intensive

Reviewer: ```suggestion
=== CPU-Intensive Thread Pools
```

Reviewer: ditto re. caps vs. hyphenated modifier; term list

> CPU Intensive is also a small thread pool (2 threads per available core by default),

Reviewer: ```suggestion
CPU Intensive is also a small thread pool (again, two threads per available core by default),
```

> but it provides a queue for accepting more tasks.

Reviewer: Anything more to say about this, compared with Light?

> === IO

Reviewer: ```suggestion
=== I/O Thread Pools
```

> IO is an elastic thread pool that grows as needed.

Reviewer: ```suggestion
The I/O thread pool is elastic, growing as needed.
```
OR
```suggestion
Unlike the other two thread pool types, the I/O type can be of any size, growing as needed.
```

> Tasks running in this pool should spend most of their time `WAITING` or `BLOCKED`

Reviewer: ```suggestion
Tasks running in this pool spend most of their time in `WAITING` or `BLOCKED` states
```

> instead of doing CPU work, so they do not compete with the other pools.

Reviewer: ```suggestion
instead of performing CPU processes, so that they do not compete with the work of the other two, more agile thread pools.
```

> Also, when a transaction is active (since many transaction managers require all

Reviewer: ```suggestion
Transaction managers also use this thread when they want all of the steps of one active transaction to be performed in a single thread. 
```

steps of the same transaction to be handled by the same thread), the `IO` pool

> is used.

Reviewer: omit per above

> === Custom Pools

Reviewer: ```suggestion
=== Custom Thread Pools
```

> Apart from the three core thread pools, Mule or some components might create additional pools for specific purposes, such as:

Reviewer: ```suggestion
Apart from the three core thread pools, Mule or some components might create additional pools for specific purposes:
```

> * NIO Selectors +

Reviewer: ```suggestion
* NIO Selector pools +
```

> Enables non-blocking IO. Each connector can create as many as required.

Reviewer: ```suggestion
Enable nonblocking I/O. Each connector can create as many as required.
```

* Recurring tasks pools +

> Some connectors or components might create specific pools to perform recurring tasks (e.g: expiration monitors, queue consumers, etc.)

Reviewer: ```suggestion
Some connectors or components (expiration monitors, queue consumers, and so on) might create specific pools to perform recurring tasks.
```

> === Proactor Pattern

Reviewer: At this level, this topic and those following are peer topics to pool types and subtopics of "Threading": is that logical and correct?

Reviewer: Also, should this title be more specific, such as Asynchronous Execution Using the Proactor Design Pattern?

> Proactor is a design pattern for asynchronous execution. To understand how the Proactor design pattern works, visit https://en.wikipedia.org/wiki/Proactor_pattern

Reviewer: ```suggestion
Proactor is a design pattern for asynchronous execution. To understand how the Proactor design pattern works, visit https://en.wikipedia.org/wiki/Proactor_pattern.
```

Reviewer: I wonder if we really need this cross-ref right at the beginning of this topic, since we subsequently define the relevant info about how this works in this context?
Consider moving to the end of the topic, introducing with something like "For more information...".

> According to this design pattern, all tasks are classified in categories that correspond to each of Mule thread pools, and each task is submitted for execution to its corresponding thread pool.

Reviewer: ```suggestion
According to this design pattern, all tasks are classified in categories that correspond to each of the Mule thread pools, and each task is submitted for execution to its corresponding thread pool.
```

Reviewer: Are these tasks and pools or processors and pools?

> Consider the following example where a flow pulls down a JSON array of Person objects described in JSON format, pushes the content through an HTTP request, picks the name of the first entry and does some additional processing:

Reviewer: ```suggestion
Consider the following example, in which a flow pulls down a JSON array of Person objects described in JSON format, pushes the content through an HTTP request, selects the name of the first entry, and performs some additional processing:
```
"pulls down"?

[source, xml, linenums]
----
<flow>
  <sftp:read path="personsArray.json" /> <1>
  <http:request path="/persons" method="POST" /> <2>
  <set-variable variableName="firstEntryName" value="#[payload[0].name]" /> <3>
  <ee:transform ... /> <4>
  <logger message="#[vars.firstEntryName]" /> <5>
</flow>
----

According to the Proactor pattern, Mule submits the tasks as follows:

> <1> The blocking operation(`<sftp:read>`) executes in the IO pool.

Reviewer: ```suggestion
<1> The blocking operation (`<sftp:read>`) executes in the I/O pool.
```

> <2> `<http:request>` is a non blocking operation. Mule performs the request in the current thread. When the flow receives the response, Mule switches to the `CPU_LIGHT` pool.

Reviewer: ```suggestion
<2> Mule performs the nonblocking operation (`<http:request>`) in the current thread. When the flow receives the response, Mule switches to the `CPU_LIGHT` pool.
```
Should the second sentence actually be <3>?

> <3> `<set-variable>` operations should be fairly quick and stay in `CPU_LIGHT`. There is no thread switch.

Reviewer: ```suggestion
<3> `<set-variable>` operations are fairly quick and stay in the `CPU_LIGHT` pool, without a thread switch.
```

> <4> `<ee:transform>`, this is potentially a computational heavy transformation, so Mule switches to the `CPU_INTENSIVE` pool.

Reviewer: ```suggestion
<4> `<ee:transform>` is potentially a computationally heavy transformation, so Mule switches to the `CPU_INTENSIVE` pool.
```
Do we want to say it switches pools or threads? The previous step refers to switching threads....

> <5> Logger stays on `CPU_INTENSIVE`. There is no thread switch. +

Reviewer: ```suggestion
<5> Logger is classified `CPU_INTENSIVE`. There is no thread switch. +
```

[NOTE]

> Due to optimizations regarding latency, thread switches are omitted when an IO or `CPU_INTENSIVE` task is followed by a `CPU_LIGHT` one. The reasoning behind this optimization is that executing said `CPU_LIGHT` task is most likely cheaper than the thread switch.

Reviewer: ```suggestion
Due to latency optimizations, thread switches are omitted when an I/O or `CPU_INTENSIVE` task is followed by a `CPU_LIGHT` task, because executing the `CPU_LIGHT` task is most likely cheaper than performing a thread switch.
```

> == Configuration

Reviewer: ```suggestion
== Configuration at the Application Level
```

> Mule runtime engine automatically configures the thread pools at startup time, by applying formulas that consider available resources such as CPU and memory. These formulas can be modified by editing the `MULE_HOME/conf/schedulers-pools.conf` file. However, MuleSoft doesn’t recommend changing the thread pools' configuration.

Reviewer: ```suggestion
Mule runtime engine automatically configures thread pools for the entire Mule instance at startup, applying formulas that consider available resources such as CPU and memory. Although you can modify these global formulas by editing the `MULE_HOME/conf/schedulers-pools.conf` file, MuleSoft doesn’t recommend it.
```

> 
> Note that the configuration is global and affects the entire Mule instance.
> 
> === Configuration at the Application Level
> 
> You can define the configuration of each pool and use it in a specific application by adding the following to your application code:

Reviewer: ```suggestion
However, you can define the configuration of each pool at the application level by adding the following to your application code:
```

[source,xml,linenums]
----
<ee:scheduler-pools gracefulShutdownTimeout="15000">
   <ee:cpu-light
           poolSize="2"
           queueSize="1024"/>
   <ee:io
           corePoolSize="1"
           maxPoolSize="2"
           queueSize="0"
           keepAlive="30000"/>
   <ee:cpu-intensive
           poolSize="4"
           queueSize="2048"/>
</ee:scheduler-pools>
----

> ==== Considerations

Reviewer: Suggest omitting. This text follows the example nicely, without needing a (solitary) subheading.

> Applying pool configurations at the application level causes Mule to create a completely new set of thread pools for the Mule app. This configuration does not change the default settings configured in the `scheduler-conf.properties` file, which is particularly important for on-premises deployments because many Mule apps can be deployed to the same Mule instance.

Reviewer: ```suggestion
Applying pool configurations at the application level causes Mule to create a completely new set of thread pools for the Mule app. This configuration does not change the default settings configured in the `scheduler-conf.properties` file, which is particularly important for on-premises deployments in which many Mule apps can be deployed to the same Mule instance.
```

> Mulesoft recommends running mule using the default settings. If you use custom thread pool configurations, MuleSoft recommends that you perform load and stress testing with all applications involved in real-life scenarios to validate any change in the threading configurations and to understand how the thread pools work in Mule 4.

Reviewer: ```suggestion
Because Mulesoft recommends running Mule using the default settings, if you decide to customize thread pool configurations, be sure to perform load and stress testing with all applications, using real-life scenarios, and consult with MuleSoft Support before moving to production using customized settings.
```

> In any case, MuleSoft recommends consulting with Support before going into production with customized settings.

Reviewer: ```suggestion

```

[[backpressure]]

> == Back-pressure

Reviewer: ```suggestion
== Back-Pressure Management
```

> Under heavy load, Mule might not have resources available to handle a specific event. This issue might occur because all threads are busy and cannot perform the handoff of the newly arrived event or because the current flow’s `maxConcurrency` value has been exceeded already.

Reviewer: ```suggestion
Back pressure can occur when, under heavy load, Mule does not have resources available to process a specific event. This issue might occur because all threads are busy and cannot perform the handoff of the newly arrived event, or because the current flow’s `maxConcurrency` value has been exceeded already.
```

> In case Mule cannot handle an event, a message is logged about the condition: `Flow 'flowName' is unable to accept new events at this time`. Also, Mule sends a notification to the flow source to perform any required actions.

Reviewer: ```suggestion
If Mule cannot handle an event, it logs the condition with the message `Flow 'flowName' is unable to accept new events at this time`. Mule also notifies the flow source, to perform any required actions.
```

> The actions Mule performs on back-pressure are specific to each connector’s source. For example, an `http:listener` might return a `503` error code, while a message broker listener might provide the option to either wait for resources to be available or drop the message.

Reviewer: ```suggestion
The actions that Mule performs as a result of back-pressure are specific to each connector’s source. For example, `http:listener` might return a `503` error code, while a message-broker listener might provide the option to either wait for resources to be available or drop the message.
```

> In some cases, a source might disconnect from a remote system to avoid getting more data than it can process and then reconnect once the server state is normalized.

Reviewer: ```suggestion
In some cases, a source might disconnect from a remote system to avoid getting more data than it can process and then reconnect after the server state is normalized.
```
